### PR TITLE
feat: API for deals tracked per miner/client

### PR DESCRIPTION
**Pull request by bajtos**
Links:
- https://github.com/filecoin-station/spark-api/pull/388.
- https://github.com/filecoin-station/spark-api/pull/386
- https://github.com/filecoin-station/spark/issues/78

Out of the scope of this pull request:

- querying deals by allocator id

Example request:

```
GET /miner/f0814049/deals/eligible/summary
```

Example response:

```json
{
  "minerId": "f0814049",
  "dealCount": 10006,
  "clients": [
    { "clientId": "f02516933", "dealCount": 6880 },
    { "clientId": "f02833886", "dealCount": 3126 }
  ]
}
```

Example request:

```
GET /clients/f0215074/deals/eligible/summary
```

Example response:

```json
{
  "clientId": "f0215074",
  "dealCount": 5350,
  "providers": [
    { "minerId": "f0406478", "dealCount": 4592 },
    { "minerId": "f0814049", "dealCount": 758 }
  ]
}
```

### CODE REVIEW

File: db/package.json [hunk 1]
```diff
@@ -14,6 +14,7 @@
   "dependencies": {
     "pg": "^8.12.0",
     "postgrator": "^7.2.0",
+    "spark-api": "https://github.com/filecoin-station/spark-api/archive/3de5c3325ef6cc40df02769b96957d33279d38c1.tar.gz",
     "spark-evaluate": "filecoin-station/spark-evaluate#main"
```
Comment (bajtos) [inline db/package.json:18]: FYI: I cannot use the git/GitHub shorthand for `spark-api` because spark-api has workspaces, and npm refuses to install the entire project when running `npm ci` limited to a single spark-stats workspace. I think it's a bug in npm, but 🤷🏻 

```
$ NPM_CONFIG_WORKSPACE=stats npm ci
(...)

npm error code 1
npm error git dep preparation failed
npm error command /opt/hostedtoolcache/node/20.16.0/x64/bin/node /opt/hostedtoolcache/node/20.16.0/x64/lib/node_modules/npm/bin/npm-cli.js install --force --cache=/home/runner/.npm --prefer-offline=false --prefer-online=false --offline=false --no-progress --no-save --no-audit --include=dev --include=peer --include=optional --no-package-lock-only --no-dry-run
npm error npm warn using --force Recommended protections disabled.
npm error npm error No workspaces found:
npm error npm error   --workspace=stats
```

**Comment by bajtos** [comment]
@juliangruber hindsight please 🙏🏻

File: README.md [hunk 2]
```diff
@@ -48,9 +56,6 @@ Base URL: http://stats.filspark.com/
 
   http://stats.filspark.com/participant//0x000000000000000000000000000000000000dEaD/reward-transfers
 
-- `GET /miners/retrieval-success-rate/summary?from=<day>&to=<day>`
-
-  http://stats.filspark.com/miners/retrieval-success-rate/summary
```
Comment (juliangruber) [inline README.md:53]: why was this removed?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -25,9 +25,11 @@ jobs:
     env:
       DATABASE_URL: postgres://postgres:postgres@localhost:5432/spark_stats
       EVALUATE_DB_URL: postgres://postgres:postgres@localhost:5432/spark_evaluate
+      API_DB_URL: postgres://postgres:postgres@localhost:5432/spark
```
Comment (juliangruber) [inline .github/workflows/ci.yml:28]: I'm strongly against this design. I thought we had settled on an architecture where services don't directly query databases not owned by them. Now we're introducing one more of this antipattern (as agreed by us).

I think the way to go instead is for spark-api to expose the data that spark-stats needs via HTTP.

**Review by juliangruber — COMMENTED** [review]
I have concerns with this PR

File: README.md (hunk 2 above)
Comment (bajtos) [reply README.md:53]: Hmm, this looks like a mistake; thanks for flagging it!

File: .github/workflows/ci.yml (hunk 3 above)
Comment (bajtos) [reply .github/workflows/ci.yml:28]: I see your concern! I actually started to implement the spark-api HTTP endpoint here: https://github.com/filecoin-station/spark-api/pull/388, but then I realised we want to put this behind Cloudflare and protect it using the same access token that we will use for other spark-stats endpoints.

@juliangruber 
How would you propose implementing your suggestion? Should spark-stats simply proxy requests for the new endpoints to spark-api?

```js
const getRetrievableDealsForMiner = async (_req, res, minerId) => {
  const apiRes = await fetch(`${sparkApiUrl}/miner/${minerId}/deals/eligible/summary`)
  assert(apiRes.ok, 502)
  json(res, await apiRes())
}
```

I'd like to prevent people from calling spark-api's endpoint directly, so I would need to implement some sort of an authentication. I can implement a fixed access token configured via Fly secrets both in spark-api and spark-stats, let spark-stats send that token in the request and let spark-api reject requests with a missing or an incorect token.

Comment (bajtos) [reply .github/workflows/ci.yml:28]: I can also YOLO it and create a tech debt that we will need to pay once we start working on the Spark Data API product:
- Option A: implement these spark-stats endpoints as HTTP redirects to spark-api
- Option B: implement these endpoints as a proxy, but don't authenticate them - allow anybody to request this data from spark-api directly if they know about this.

Option A:

```js
const getRetrievableDealsForMiner = async (_req, res, minerId) => {
  res.setHeader(
    'location',
    `${sparkApiUrl}/miner/${minerId}/deals/eligible/summary`
  )
  res.statusCode = 302
  res.end()
}

Comment (bajtos) [reply .github/workflows/ci.yml:28]: Depending on what direction we choose, I'll need to rework the API for deals tracked per allocator (https://github.com/filecoin-station/spark-stats/pull/196) as well.

Comment (bajtos) [reply .github/workflows/ci.yml:28]: I was thinking about this a bit more.

To track DDO deals, I will need to build a new service that listens to Filecoin actor events and maintains its own deal database. Then both spark-api and spark-stats will need to get data about eligible deals from that service, most likely over HTTP. 

We may want to make that data about deals public to allow 3rd parties to verify that Spark is sampling the deals fairly.

In that light, here is my plan:

- Copy the implementation of the APIs introduced here to https://github.com/filecoin-station/spark-stats/pull/196
- Rework spark-stats endpoints to redirect to spark-api, drop the dev-dependency on spark-api repository.

Rationale:
- spark-api is already behind Cloudflare and it accepts only requests made via Cloudflare. If spark-stats implements a reverse proxy, we will create unnecessary network traffic we need to pay for (Cloudflare -> spark-stats -> Cloudflare -> spark-api).
- Since we don't know yet whether we want/need to make the data about eligible deals private, let's keep it public for now as it requires less work.

Comment (juliangruber) [reply .github/workflows/ci.yml:28]: > I was thinking about this a bit more.
> 
> To track DDO deals, I will need to build a new service that listens to Filecoin actor events and maintains its own deal database. Then both spark-api and spark-stats will need to get data about eligible deals from that service, most likely over HTTP.
> 
> We may want to make that data about deals public to allow 3rd parties to verify that Spark is sampling the deals fairly.

👍 

> 
> In that light, here is my plan:
> 
> * Copy the implementation of the APIs introduced here to [feat: API for deals tracked per allocator #196](https://github.com/filecoin-station/spark-stats/pull/196)

I don't understand this, how are you going to copy APIs to a closed PR?

> * Rework spark-stats endpoints to redirect to spark-api, drop the dev-dependency on spark-api repository.
> 
> Rationale:
> 
> * spark-api is already behind Cloudflare and it accepts only requests made via Cloudflare. If spark-stats implements a reverse proxy, we will create unnecessary network traffic we need to pay for (Cloudflare -> spark-stats -> Cloudflare -> spark-api).
> * Since we don't know yet whether we want/need to make the data about eligible deals private, let's keep it public for now as it requires less work.

Redirect also works for me, if it's a temporary one (302). We don't want anyone to remember spark-api's URL if it's not an API we want to commit to hosting

Comment (bajtos) [reply .github/workflows/ci.yml:28]: > > Copy the implementation of the APIs introduced here to https://github.com/filecoin-station/spark-stats/pull/196
> I don't understand this, how are you going to copy APIs to a closed PR?

Sorry, I should have re-read the comment before submitting.

I want to copy the API endpoints from this PR and #196 to spark-api.

Comment (bajtos) [reply .github/workflows/ci.yml:28]: > Redirect also works for me, if it's a temporary one (302). We don't want anyone to remember spark-api's URL if it's not an API we want to commit to hosting

💯

Comment (juliangruber) [reply .github/workflows/ci.yml:28]: Sounds good!

Comment (bajtos) [reply .github/workflows/ci.yml:28]: spark-api PR:
- https://github.com/filecoin-station/spark-api/issues/404

File: README.md (hunk 2 above)
Comment (bajtos) [reply README.md:53]: Found it. This was a duplicate entry, the endpoint is already documented on lines 19-21 above.

Screenshot 2024-09-02 at 12 12 33
